src/instruction.rs

```diff
@@ -1080,19 +1131,91 @@ mod test_instruction_display {
 #[derive(Clone, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
 pub enum Qubit {
     Fixed(u64),
+    #[serde(skip)] // todo
+    Placeholder(QubitPlaceholder),
     Variable(String),
 }
 
+impl Qubit {
+    fn resolve_placeholder<R>(&mut self, resolver: R)
+    where
+        R: Fn(&QubitPlaceholder) -> Option<u64>,
+    {
+        if let Qubit::Placeholder(placeholder) = self {
+            if let Some(resolved) = resolver(placeholder) {
+                *self = Qubit::Fixed(resolved);
+            }
+        }
+    }
+
+    // For comparison, this takes a more active approach than the above, in that it can allocate new indices
+    // using the placeholder. This seems unnecessary and so I've commented it out, but leaving it here
+    // for commentary while in draft
+    // fn resolve_placeholder<I, R>(
+    //     &mut self,
+    //     indices: &mut I,
+    //     resolver: R,
+    // ) -> Result<Option<(QubitPlaceholder, u64)>, ()>
+    // where
+    //     I: Iterator<Item = u64>,
+    //     R: Fn(QubitPlaceholder) -> Option<u64>,
+    // {
+    //     if let Qubit::Placeholder(placeholder) = self {
+    //         if let Some(resolved) = resolver(placeholder.clone()) {
+    //             *self = Qubit::Fixed(resolved);
+    //             Ok(None)
+    //         } else if let Some(next) = indices.next() {
+    //             let return_value = Some((placeholder.clone(), next));
+    //             *self = Qubit::Fixed(next);
+    //             Ok(return_value)
+    //         } else {
+    //             Err(())
+    //         }
+    //     } else {
+    //         Ok(None)
+    //     }
+    // }
+}
+
 impl fmt::Display for Qubit {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         use Qubit::*;
         match self {
             Fixed(value) => write!(f, "{}", value),
+            Placeholder(_) => todo!(),
```

Comment:
### Option 1 - Break the expectation that Display outputs Quil

This mirrors what `pyQuil` does today, and I'm assuming would give rise to something like a `Quil` trait with a `to_quil() -> Result<String, Err>`. I like how the trait would let us implement a common interface for a `Program` and each of the instruction types.

### Option 2 - Resolve at Display time

Probably the quickest approach in terms of development time. I'm not sure if I love having to bake in "magic" like that in a Display function, but since it would have to clone the `Program` at least it wouldn't unexpectedly mutate the object. The cloning itself isn't fantastic though.

### Option 3 - Introduce a `ParserProgram`

I would prefer Option 1 to this. Mainly because a `ParserProgram` only solves this at the program level and leaves us with the same problem for each of the `Instruction` types. The parser can/should only parse valid Quil so outside of a bug, I don't see how it would build a `Program` that couldn't be turned back into valid Quil. If I remember correctly, the parser parses `Instruction`s, not a `Program`, so a `ParserProgram` wouldn't help guarantee that the parser couldn't mistakenly use something like a program either.

Reply:
## Option 1

Agreed, that could work out nicely - the question is then what the standard is for `Display` and thus `to_string()`, if we even still implement those.

## Option 2

We may be able to avoid the clone by resolving each instruction as it's displayed, so instructions that don't require any resolution aren't mutated or copied.

## Option 3

Ah, to clarify, the `ParserProgram` would contain`ParserInstruction`s, which would be infallibly and unambiguously represented in Quil. @Shadow53 suggested the name `QuilProgram` and thus `QuilInstruction` instead.

It might look like

```rust
struct QuilProgram(Vec<QuilInstruction>); // Since we don't need to keep track of DECLARE separately from PULSE, etc
```

It's a simpler wrapper that may not actually need to be public, if I'm thinking about it right. I don't think this change would require any changes in pyQuil; it would be a tool internal to the library and the bindings.

Reply:
I am leaning towards option 1, and removing the `Display` implementation altogether (replacing it with a `to_quil()` method which fails if there are unresolved placeholders).

Agreed option 2 feels like potentially unexpected behaviour.

Option 3 sounds nice because types are nice, but I don't think it is as simple/clean as option 1 without providing more.

Reply:
Okay, I'm on board, if we do it as @notmgsk says here - remove `Display` altogether, which is breaking but prevents confusion. There should also likely be a `from_quil` trait which does the parsing, rather than `FromStr`.

Reply:
I went with ⬆️ that approach. I'm a little unsettled with it, namely in that now a `Placeholder` can be everywhere a `Qubit` is, even when that is not possibly the case (such as when it's being parsed from a string). That means that there are places that have to account for that not-possible possibility; `unreachable!` or `expect` are options of course but I'd like to avoid those as well.

I put two methods on the `Quil` trait:

* `to_quil` is fallible and if it succeeds, it outputs something `Display`.
* `to_quil_or_debug` is available if the struct implements `Debug` and falls back to that representation if Quil stringification fails. This feels better to me than a panic and works well for the tests.